File: hooks/desktop-control/desktop_controller.js
// Desktop Controller - Node.js wrapper for Python PyAutoGUI bridge
const { spawn } = require('child_process');
const path = require('path');

class DesktopController {
  constructor(options = {}) {
    this.failsafe = options.failsafe !== false;
    this.requireApproval = options.requireApproval || false;
    this.pythonProcess = null;
    this.initialized = false;
    this.commandQueue = [];
    this.requestId = 0;
    this.pendingRequests = new Map();
    
    this._init();
  }
  
  async _init() {
    const pythonScript = path.join(__dirname, 'desktop_controller.py');
    
    this.pythonProcess = spawn('python', [pythonScript], {
      stdio: ['pipe', 'pipe', 'pipe']
    });
    
    this.pythonProcess.stdout.on('data', (data) => {
      const lines = data.toString().trim().split('\n');
      for (const line of lines) {
        if (line) {
          try {
            const response = JSON.parse(line);
            if (response.requestId && this.pendingRequests.has(response.requestId)) {
              const { resolve, reject } = this.pendingRequests.get(response.requestId);
              this.pendingRequests.delete(response.requestId);
              
              if (response.success) {
                resolve(response.result);
              } else {
                reject(new Error(response.error));
              }
            }
          } catch (e) {
            console.error('Invalid response:', line);
          }
        }
      }
    });
    
    this.pythonProcess.stderr.on('data', (data) => {
      console.error('Python error:', data.toString());
    });
    
    this.pythonProcess.on('close', (code) => {
      console.log(`Python process exited with code ${code}`);
      this.initialized = false;
    });
    
    this.initialized = true;
  }
  
  async _sendCommand(method, args = [], kwargs = {}) {
    if (!this.initialized) {
      throw new Error('Desktop controller not initialized');
    }
    
    return new Promise((resolve, reject) => {
      const requestId = ++this.requestId;
      this.pendingRequests.set(requestId, { resolve, reject });
      
      const command = JSON.stringify({
        requestId,
        method,
        args,
        kwargs
      });
      
      this.pythonProcess.stdin.write(command + '\n');
      
      // Timeout after 30 seconds
      setTimeout(() => {
        if (this.pendingRequests.has(requestId)) {
          this.pendingRequests.delete(requestId);
          reject(new Error('Command timeout'));
        }
      }, 30000);
    });
  }
  
  // Mouse functions
  async moveMouse(x, y, options = {}) {
    const duration = options.duration || 0;
    return this._sendCommand('move_mouse', [x, y], { duration });
  }
  
  async moveRelative(xOffset, yOffset, duration = 0) {
    return this._sendCommand('move_relative', [xOffset, yOffset], { duration });
  }
  
  async click(x, y, options = {}) {
    const button = options.button || 'left';
    const clicks = options.clicks || 1;
    
    if (x !== undefined && y !== undefined) {
      return this._sendCommand('click', [x, y], { button, clicks });
    }
    return this._sendCommand('click', [], { button, clicks });
  }
  
  async drag(startX, startY, endX, endY, duration = 0.5) {
    return this._sendCommand('drag', [startX, startY, endX, endY], { duration });
  }
  
  async scroll(clicks, direction = 'vertical') {
    return this._sendCommand('scroll', [clicks], { direction });
  }
  
  async getMousePosition() {
    const result = await this._sendCommand('get_mouse_position');
    return { x: result.x, y: result.y };
  }
  
  // Keyboard functions
  async typeText(text, options = {}) {
    const interval = options.interval || 0;
    const wpm = options.wpm || null;
    return this._sendCommand('type_text', [text], { interval, wpm });
  }
  
  async press(key, presses = 1) {
    return this._sendCommand('press', [key], { presses });
  }
  
  async hotkey(...keys) {
    return this._sendCommand('hotkey', keys);
  }
  
  async keyDown(key) {
    return this._sendCommand('key_down', [key]);
  }
  
  async keyUp(key) {
    return this._sendCommand('key_up', [key]);
  }
  
  // Screen functions
  async screenshot(options = {}) {
    const region = options.region || null;
    return this._sendCommand('screenshot', [], { region });
  }
  
  async getPixelColor(x, y) {
    const result = await this._sendCommand('get_pixel_color', [x, y]);
    return { r: result.r, g: result.g, b: result.b };
  }
  
  async getScreenSize() {
    const result = await this._sendCommand('get_screen_size');
    return { width: result.width, height: result.height };
  }
  
  async findOnScreen(imagePath, confidence = 0.8) {
    return this._sendCommand('find_on_screen', [imagePath], { confidence });
  }
  
  // Window functions
  async getAllWindows() {
    return this._sendCommand('get_all_windows');
  }
  
  async activateWindow(titleSubstring) {
    return this._sendCommand('activate_window', [titleSubstring]);
  }
  
  // Cleanup
  dispose() {
    if (this.pythonProcess) {
      this.pythonProcess.kill();
      this.pythonProcess = null;
      this.initialized = false;
    }
  }
}

module.exports = { DesktopController };

File: hooks/desktop-control/desktop_controller.py
# Desktop Controller - Python bridge for PyAutoGUI
import pyautogui
import json
import sys
import os

class DesktopController:
    def __init__(self, failsafe=True, requireApproval=False):
        pyautogui.FAILSAFE = failsafe
        self.requireApproval = requireApproval
        
    def move_mouse(self, x, y, duration=0):
        """Move mouse to absolute coordinates"""
        if self.requireApproval:
            if not self._confirm(f"Move mouse to ({x}, {y})?"):
                return False
        pyautogui.moveTo(x, y, duration=duration)
        return True
        
    def move_relative(self, x_offset, y_offset, duration=0):
        """Move mouse relative to current position"""
        x, y = pyautogui.position()
        pyautogui.moveTo(x + x_offset, y + y_offset, duration=duration)
        return True
        
    def click(self, x=None, y=None, button='left', clicks=1):
        """Perform mouse click"""
        if x is not None and y is not None:
            pyautogui.click(x, y, clicks=clicks, button=button)
        else:
            pyautogui.click(clicks=clicks, button=button)
        return True
        
    def drag(self, start_x, start_y, end_x, end_y, duration=0.5, button='left'):
        """Drag and drop operation"""
        pyautogui.moveTo(start_x, start_y)
        pyautogui.dragTo(end_x, end_y, duration=duration, button=button)
        return True
        
    def scroll(self, clicks, direction='vertical'):
        """Scroll mouse wheel"""
        if direction == 'vertical':
            pyautogui.scroll(clicks)
        else:
            pyautogui.hscroll(clicks)
        return True
        
    def get_mouse_position(self):
        """Get current mouse coordinates"""
        x, y = pyautogui.position()
        return {'x': x, 'y': y}
        
    def type_text(self, text, interval=0, wpm=None):
        """Type text with configurable speed"""
        if wpm:
            # Calculate interval from WPM (average 5 chars per word)
            interval = 60 / (wpm * 5)
        pyautogui.typewrite(text, interval=interval)
        return True
        
    def press(self, key, presses=1):
        """Press and release a key"""
        pyautogui.press(key, presses=presses)
        return True
        
    def hotkey(self, *keys):
        """Execute keyboard shortcut"""
        pyautogui.hotkey(*keys)
        return True
        
    def key_down(self, key):
        """Hold key down"""
        pyautogui.keyDown(key)
        return True
        
    def key_up(self, key):
        """Release key"""
        pyautogui.keyUp(key)
        return True
        
    def screenshot(self, region=None):
        """Capture screen or region"""
        from PIL import Image
        screenshot = pyautogui.screenshot(region=region)
        return screenshot
        
    def get_pixel_color(self, x, y):
        """Get color of pixel at coordinates"""
        r, g, b = pyautogui.pixel(x, y)
        return {'r': r, 'g': g, 'b': b}
        
    def get_screen_size(self):
        """Get screen resolution"""
        width, height = pyautogui.size()
        return {'width': width, 'height': height}
        
    def find_on_screen(self, image_path, confidence=0.8):
        """Find image on screen (requires OpenCV)"""
        try:
            location = pyautogui.locateOnScreen(image_path, confidence=confidence)
            if location:
                return {
                    'x': location.left,
                    'y': location.top,
                    'width': location.width,
                    'height': location.height,
                    'center_x': location.left + location.width // 2,
                    'center_y': location.top + location.height // 2
                }
            return None
        except Exception as e:
            return {'error': str(e)}
            
    def get_all_windows(self):
        """List all open windows (Windows only)"""
        try:
            import pygetwindow as gw
            windows = gw.getAllWindows()
            return [{'title': w.title, 'left': w.left, 'top': w.top, 
                    'width': w.width, 'height': w.height} for w in windows if w.title]
        except:
            return [{'error': 'Window management requires pygetwindow'}]
            
    def activate_window(self, title_substring):
        """Bring window to front"""
        try:
            import pygetwindow as gw
            window = gw.getWindowsWithTitle(title_substring)[0]
            window.activate()
            return True
        except Exception as e:
            return {'error': str(e)}
            
    def _confirm(self, action):
        """Ask for user confirmation"""
        response = input(f"Allow: {action} [y/n]: ")
        return response.lower() == 'y'

# Command-line interface for Node.js bridge
def main():
    dc = DesktopController()
    
    while True:
        try:
            line = input()
            if not line:
                continue
                
            cmd = json.loads(line)
            method = cmd.get('method')
            args = cmd.get('args', [])
            kwargs = cmd.get('kwargs', {})
            
            if hasattr(dc, method):
                result = getattr(dc, method)(*args, **kwargs)
                print(json.dumps({'success': True, 'result': result}))
            else:
                print(json.dumps({'success': False, 'error': f'Method {method} not found'}))
                
        except json.JSONDecodeError as e:
            print(json.dumps({'success': False, 'error': f'Invalid JSON: {str(e)}'}))
        except Exception as e:
            print(json.dumps({'success': False, 'error': str(e)}))
        
        sys.stdout.flush()

if __name__ == '__main__':
    main()